[Bug]: Debug zppy diffs for v3.0.0

tomvothecoder opened the pull request:
## Description

<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

### Summary of Changes
- Closes #930 
- Fix RMSE and CORR text position

### TLDR: Conclusion

I'm confident that v3.0.0 is working correctly since the test (model) and reference (observation) files align. The differences in the "Model - Observation" plots are expected because:

- **Subtle Regridding Differences:** xESMF and ESMF have minor differences that lead to slight variations in regridded data.
- **Floating-Point Sensitivity:** Comparing floating-point differences can amplify small discrepancies.
- **Minor Data Variations:** Even slight differences (e.g., mismatched NaNs) in the test or reference files can result in significant changes in `diff.nc`.
- **Difference Calculation:** Since `diff.nc` is computed as test minus reference, any small discrepancy can produce large differences, which then appear in the plot.


### Debugging Overview
Some "Model - Observation" subplots show large differences between E3SM Diags v2.12.1 and v3.0.0rc2. These plots use the `diff.nc` file (test - ref). We need to determine why the plot differences are so large in some cases.

#### 1. First, let's compare the output files for test (`test.nc`) and reference (`ref.nc`) to see if these align between `e3sm_diags` versions. 

Source: [`regression_nc.ipynb`](https://github.com/E3SM-Project/e3sm_diags/blob/bug/930-zppy-diffs/auxiliary_tools/cdat_regression_testing/930-zppy-diffs/v2/regression_nc.ipynb). With relative tolerance of 1e-5 (`rol=1e-5`): 

- 30/34 matching files
- 3/34 mismatch errors due to ccb regional subsetting differences, these are expected
  - Affects `ERA5-TREFHT-land`, `MERRA2-TAUXY-land`, and `MERRA2-TREFHT-land`
  - Related to https://github.com/E3SM-Project/e3sm_diags/pull/794
- 1/34 not equal files -- number of different elements are really small and stats (min, max, mean, sum) are similar (MERRA2-OMEGA below)
  ```python
  Comparing:
      * /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5-OMEGA-850-ANN-global_test.nc
      * /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5-OMEGA-850-ANN-global_test.nc
      * var_key: OMEGA

  Not equal to tolerance rtol=0.0001, atol=0

  Mismatched elements: 17 / 64800 (0.0262%)
  Max absolute difference among violations: 2.15263471e-05
  Max relative difference among violations: 0.00060596
  ACTUAL: array([[      nan,       nan,       nan, ...,       nan,       nan,
                nan],
        [      nan,       nan,       nan, ...,       nan,       nan,...
  DESIRED: array([[      nan,       nan,       nan, ...,       nan,       nan,
                nan],
        [      nan,       nan,       nan, ...,       nan,       nan,...
  ```
  ```python
  Checking variable OMEGA
  Dev Path: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/MERRA2/MERRA2-OMEGA-850-ANN-global_test.nc
  Main Path: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/MERRA2/MERRA2-OMEGA-850-ANN-global_test.nc
  -------------------------------------
  * Min - dev: -248.597013, main: -248.597015
  * Max - dev: 604.741699, main: 604.741211
  * Mean - dev: 0.446415, main: 0.446415
      * Absolute Mean Diff: 1.3810975812988957e-07
      * Relative Mean Diff: 0.000031%
  * Sum - dev: 26022.445543, main: 26022.439453
      * Absolute Sum Diff: 0.006090177481382852
      * Relative Sum Diff: 0.000023%
  ```

##### Great news! The test and reference files are within tolerance, so the first two of three plots show no image diffs. Now, we need to analyze `diffs.nc` for differences in the third plot, "Model-Observation."

#### I asked ChatGPT two questions:

##### Can floating point comparison between the differences of two floating point arrays result in large differences? 

> Yes, floating-point comparison between the **differences** of two floating-point arrays can result in large differences due to the following reasons:  
> 
> 1. **Loss of Precision** – Floating-point arithmetic is not exact, and small rounding errors can accumulate when computing differences.  
> 2. **Subtraction Amplifies Error** – If two nearly equal floating-point numbers are subtracted, the relative error in the result can be significantly larger than the original numbers.  
> 3. **NaN and Masking Effects** – Mismatched NaNs, infinities, or slight variations in masked values can cause major differences in computed differences.  
> 4. **Chaining of Differences** – If `A - B` and `C - D` have small floating-point errors individually, `(A - B) - (C - D)` can exaggerate those errors, leading to large variations.  
> 
> Thus, comparing differences of floating-point arrays is unreliable, especially when precision errors and masking differences are involved.

##### I have a plot called "Model - Observation" which plots the difference of two floating point arrays (model and observation). If I compare the difference in the plots between two branches (main vs. development), should I expect large difference due to shifting nan positions from regridding even if nan counts are the same?

> The presence of `nan` values themselves doesn’t directly cause large numerical differences in the subtraction (i.e., `model - observation`). However, if regridding causes the positions of these `nan` values to shift—even if the total count remains the same—it can lead to differences in the resulting "Model - Observation" plot. Here’s why:
> 
> 1. **Local Data Differences:**  
>    When `nan` values shift positions between branches, the set of valid (non-`nan`) data points being subtracted at each grid location can change. Even if the overall count of `nan` values remains the same, their spatial redistribution may result in different local comparisons. For instance, a cell that is valid in one branch might be `nan` in the other, and vice versa.
> 
> 2. **Impact on Interpolation/Regridding:**  
>    Regridding often involves interpolation or other spatial transformations. If the algorithm encounters `nan` values in different positions, the interpolation might use a different set of neighboring points, leading to subtle or even noticeable differences in the computed values at certain grid points.
> 
> 3. **Visual Interpretation:**  
>    While the underlying arithmetic operation (subtraction) is straightforward, the visualization can be sensitive to how these differences manifest spatially. Shifting `nan` positions can change the appearance of the plot, especially near boundaries or regions with steep gradients, even if the overall error statistics (like mean or total error) remain similar.
> 
> 4. **Consistency Check:**  
>    It’s a good practice to compare the actual arrays (both model and observation) from both branches to see if there are differences beyond just the location of `nan` values. If the valid numerical data are nearly identical but the `nan` positions differ, then the plot differences are likely due to how the data are being masked or interpolated during regridding rather than a flaw in the subtraction itself.
> 
> **In summary:**  
> You should not expect large differences solely due to floating point subtraction. However, if regridding shifts the positions of `nan` values between the branches, this can lead to changes in the computed differences at specific grid points and alter the appearance of the plot. The differences are a result of the spatial distribution of valid versus invalid data rather than the arithmetic operation itself.

#### These explanations reinforce that differences are expected in some "Model-Observation" plots.

#### Comparing `diff.nc` between codebases is unreliable, but let's still group image diffs by suspected or known causes.

##### 1. Expected due to bug found on old CDAT codebase ([here](https://github.com/E3SM-Project/e3sm_diags/pull/794))

Affected variables:
- [x] MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-ANN-global.png  
- [x] MISRCOSP-CLDLOW_TAU1.3_MISR-ANN-global.png      
- [x] MISRCOSP-CLDLOW_TAU9.4_MISR-ANN-global.png

##### 2. Expected due to RMSE/CORR positioning bug (fixed in this PR)

Affected variables:
- [x] MERRA2-PSL-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-PSL-ANN-global.png_diff.png)) -- RMSE and CORR text diff      
- [x] ceres_ebaf_surface_v4.1-ALBEDO_SRF-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/CERES-EBAF-surface-v4.1/ceres_ebaf_surface_v4.1-ALBEDO_SRF-ANN-global.png_diff.png)) -- RMSE/CORR

##### 3. Expected due to `nan` location mismatches, caused by regridding differences between xCDAT/xESMF and CDAT/ESMF when [downscaling variables](https://github.com/E3SM-Project/e3sm_diags/blob/9b92f8f6255f3f3b6b36bffb3174d0ef2061a408/e3sm_diags/driver/utils/regrid.py#L342-L428).

> Why are there subtle differences between xESMF and ESMF regridders when downscaling variables?
> 
> Subtle differences between xESMF and ESMF regridders when downscaling variables arise due to differences in their underlying algorithms and precision. While both aim to regrid data, xESMF (built on the xCDAT framework) and ESMF (Earth System Modeling Framework) may use slightly different methods for interpolation, handling of missing values, or rounding errors. These discrepancies, though minor, can propagate and result in small but noticeable variations in the regridded data.

Affected variables:
- [x] HadISST_CL-SST-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/SST_CL_HadISST/HadISST_CL-SST-ANN-global.png_diff.png))
- [x] HadISST_PD-SST-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/SST_PD_HadISST/HadISST_PD-SST-ANN-global.png_diff.png))   
- [x] HadISST_PI-SST-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/SST_PI_HadISST/HadISST_PI-SST-ANN-global.png_diff.png)) 
- [x] MACv2-AODVIS-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/AOD_550/MACv2-AODVIS-ANN-global.png_diff.png))     
- [x] MERRA2-OMEGA-850-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-OMEGA-850-ANN-global.png_diff.png))
- [x] MERRA2-T-850-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-T-850-ANN-global.png_diff.png))
- [x] MERRA2-TREFMXAV-ANN-global.png  -([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-TREFMXAV-ANN-global.png_diff.png))     
- [x] MERRA2-TREFMNAV-ANN-global.png -- ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-TREFMNAV-ANN-global.png_diff.png)) maybe also regional subsetting diff (I think due to "ccb" flag [here](https://github.com/E3SM-Project/e3sm_diags/blob/ca41b0e5d913610c88410928951f1ed11c75663f/e3sm_diags/plot/cartopy/lat_lon_plot.py#L91))


##### Expected due to land-sea mask differences between xCDAT/xESMF and CDAT/ESMF. [Details](https://github.com/E3SM-Project/e3sm_diags/blob/9b92f8f6255f3f3b6b36bffb3174d0ef2061a408/e3sm_diags/driver/utils/regrid.py#L177-L268)](https://github.com/E3SM-Project/e3sm_diags/blob/9b92f8f6255f3f3b6b36bffb3174d0ef2061a408/e3sm_diags/driver/utils/regrid.py#L177-L268).

> There seems to be a minor difference in `nan` locations for the regridded land-sea mask produced using xCDAT + xESMF vs. CDAT + ESMF (`method="bilinear"`). xCDAT has a little bit more `nan` values than CDAT. This causes `x and y nan location mismatch` when comparing array values. I don't know the exact reason why there are differences, but they seem insignificant enough to accept.
> 
> The absolute sum, mean, and nan count values are similar across variables. The diff plots also show no visible differences. All of the scripts for debugging these variables can be found [here](https://github.com/E3SM-Project/e3sm_diags/tree/refactor/792-lat-lon-debug/auxiliary_tools/cdat_regression_testing/792-lat-lon-run-script).
> -- <cite>https://github.com/E3SM-Project/e3sm_diags/pull/794</cite>

Affected variables:
- [x] ERA5-TREFHT-ANN-land.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/ERA5/ERA5-TREFHT-ANN-land.png_diff.png))
- [x] MERRA2-TAUXY-ANN-ocean.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-TAUXY-ANN-ocean.png_diff.png))      
- [x] MERRA2-TREFHT-ANN-land.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/MERRA2/MERRA2-TREFHT-ANN-land.png_diff.png))
 
##### Expected due to differences between xCDAT/xESMF vs. CDAT/ESMF when downscaling variables. These affected variables have large diffs, so I did further investigation.
 
 Affected variables:
- [x] ERA5_ext-QREFHT-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global.png_diff.png))
- [x] ERA5_ext-U10-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global.png_diff.png))  
- [x] GPCP_v3.2-PRECT-ANN-global.png ([diff](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global.png_diff.png))


Further debugging to validate expected for affected variables:

1. Compare the test and ref files within relative tolerance of 1-e5
Results: All pass
```python
Relative tolerance: 1e-05, Absolute tolerance: 0
Comparing test.nc and ref.nc files
==================================================
QREFHT - ref.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_ref.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_ref.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_ref.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_ref.nc)
Arrays are within relative tolerance.

QREFHT - test.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_test.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_test.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_test.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_test.nc)
Arrays are within relative tolerance.

U10 - ref.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_ref.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_ref.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_ref.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_ref.nc)
Arrays are within relative tolerance.

U10 - test.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_test.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_test.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_test.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_test.nc)
Arrays are within relative tolerance.

PRECT - ref.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_ref.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_ref.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_ref.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_ref.nc)
Arrays are within relative tolerance.

PRECT - test.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_test.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_test.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_test.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_test.nc)
Arrays are within relative tolerance.
```

2. Compare the `diff.nc` files used for "Model - Observation" plot
Result: Large differences

```python
Comparing diff.nc files
==================================================
QREFHT - diff.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_diff.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_diff.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_diff.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_diff.nc)

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 64795 [/](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/) 64800 (100%)
Max absolute difference among violations: 2.54303768
Max relative difference among violations: 1123.35406349
 ACTUAL: array([[0.095432, 0.09547 , 0.095508, ..., 0.142239, 0.142297, 0.142357],
       [0.116816, 0.116988, 0.117155, ..., 0.164408, 0.164593, 0.164751],
       [0.130151, 0.130483, 0.130807, ..., 0.177853, 0.178253, 0.178665],...
 DESIRED: array([[0.095825, 0.095806, 0.095784, ..., 0.142794, 0.142797, 0.142803],
       [0.117481, 0.117646, 0.117798, ..., 0.165032, 0.165229, 0.165408],
       [0.130066, 0.130319, 0.130569, ..., 0.178232, 0.178459, 0.178696],...

U10 - diff.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_diff.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_diff.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_diff.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_diff.nc)

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 64797 [/](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/) 64800 (100%)
Max absolute difference among violations: 2.318882
Max relative difference among violations: 1363.4496
 ACTUAL: array([[-1.973981, -1.982043, -1.990005, ..., -1.485713, -1.496813,
        -1.507838],
       [-1.794035, -1.803559, -1.810186, ..., -1.385356, -1.391502,...
 DESIRED: array([[-1.766862, -1.777283, -1.787618, ..., -1.275891, -1.28755 ,
        -1.29917 ],
       [-1.732177, -1.732711, -1.732202, ..., -1.345124, -1.346935,...

PRECT - diff.nc
--------------------------------------------------
Main: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_diff.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_diff.nc)
Dev: [/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_diff.nc](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_diff.nc)

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 64796 [/](https://vscode-remote+ssh-002dremote-002bchrysalis.vscode-resource.vscode-cdn.net/) 64800 (100%)
Max absolute difference among violations: 4.72772718
Max relative difference among violations: 773616.0423584
 ACTUAL: array([[-0.14976 , -0.135835, -0.128921, ..., -0.128685, -0.118462,
        -0.120224],
       [-0.053909, -0.051243, -0.051175, ..., -0.03275 , -0.023618,...
 DESIRED: array([[-0.100752, -0.087916, -0.081792, ..., -0.078111, -0.072161,
        -0.072393],
       [-0.075459, -0.068423, -0.064549, ..., -0.056744, -0.045891,...
```


3. Compare stats for `diff.nc` files 
Results -- All close enough to be comfortable (I think) 

```python
Comparing stats of diff.nc files
==================================================
QREFHT - diff.nc
--------------------------------------------------
Main: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_diff.nc
Dev: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-QREFHT-ANN-global_diff.nc
           min       max      mean           sum  nan_count
main -6.794676  5.828569  0.248934  16130.897162          0
dev  -7.093775  5.758040  0.250263  16217.108398          0
U10 - diff.nc
--------------------------------------------------
Main: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_diff.nc
Dev: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/ERA5/ERA5_ext-U10-ANN-global_diff.nc
           min       max      mean          sum  nan_count
main -5.607303  8.402979  0.035228  2282.743652          0
dev  -5.130340  7.829661  0.035636  2309.217285          0
PRECT - diff.nc
--------------------------------------------------
Main: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_diff.nc
Dev: /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-global_diff.nc
           min       max      mean          sum  nan_count
main -7.050051  8.961453  0.080120  5191.758895          0
dev  -8.213846  8.937169  0.080524  5217.875488          0
```

### Step-by-step for Debugging

1. Create a run script with a `.cfg` that isolates the variables below
2. Run the `run_script` using v2.12.1 and v3.0.0rc2
3. Compare the `.nc` differences for `test.nc` and `ref.nc` files (`regression_nc.ipynb`)
4. Compare `.nc` differences for `diff.nc` files (`regression_nc_diffs.ipynb`) 
5. Debug differences


## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

>

tomvothecoder commented on e3sm_diags/plot/utils.py:
```diff
@@ -743,7 +743,7 @@ def _add_rmse_corr_text(
 
     if left_text_pos is None:
-        left_text_pos = (0.6335, -0.0105)
+        left_text_pos = (0.6635, -0.0105)
```
Aligns with https://github.com/E3SM-Project/e3sm_diags/blob/ca41b0e5d913610c88410928951f1ed11c75663f/e3sm_diags/plot/cartopy/lat_lon_plot.py#L223

chengzhuzhang replied:
I suppose this will fix other 3 panel plots that uses the lat_lon_plot.py? because I do see same problems in other sets that has the string `RMSE` `CORR` mis-placed, some are over lapping with figures.

chengzhuzhang replied:
Okay, I checked, `lat_lon`, `zonal_mean_2d/stratosphere`, `meridional_mean_2d`, `polar` will all be fixed.

tomvothecoder commented:
@chengzhuzhang For the `OMI-MLS-TCO-ANN-60S60N` variable, we are getting mismatching shapes for the `diff.nc` file.
**Any suggestions on what we should do here? Related to #739.**

### Root cause
To determine the lower resolution for regridding, [regrid_to_lower_res()](https://github.com/E3SM-Project/e3sm_diags/blob/ca41b0e5d913610c88410928951f1ed11c75663f/e3sm_diags/driver/utils/general.py#L240-L264) in v2.12.1 compares the X axes, while [`align_grids_to_lower_res()`](https://github.com/E3SM-Project/e3sm_diags/blob/060308bf3a7d9b9c087f403ca3f7491f8f7c4fba/e3sm_diags/driver/utils/regrid.py#L342-L429) in v3.0.0 compares the "Y" axes. 



### Debugging
Test var has a shape of (120, 360) while ref var has a shape of (120, 288).

-  v2.12.1 -- shape of (120, 288) 
   - Compares the **"X" axes** to determine the lower resolution (`<=`)
   - Ref var is considered low resolution due to "X" of 288. 
   - Test var is regridded from (120, 360) to (120, 288)
- v3.0.0rc1/rc2 -- shape of (120, 360)
   - Compares the **"Y" axes** to determine the lower resolution (`<=`) 
   - Test and ref vars have equal "Y" of 120.
   - Ref var is regridded from (120, 288) to (120, 360). 

```python
Comparing:
    * /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-branch-930-zppy-diffs/lat_lon/OMI-MLS/OMI-MLS-TCO-ANN-60S60N_diff.nc
    * /lcrc/group/e3sm/public_html/cdat-migration-fy24/25-02-04-main-zppy-diffs/lat_lon/OMI-MLS/OMI-MLS-TCO-ANN-60S60N_diff.nc
    * var_key: TCO
    
Not equal to tolerance rtol=0.0001, atol=0

(shapes (120, 360), (120, 288) mismatch)
 ACTUAL: array([[4.909916, 5.552422, 6.897028, ..., 7.011181, 8.307238, 6.755878],
       [7.468704, 7.0319  , 7.010069, ..., 4.298189, 5.202745, 6.788332],
       [5.496756, 6.867319, 7.179531, ..., 5.23127 , 5.508911, 6.141758],...
 DESIRED: array([[2.213444, 1.642891, 2.503925, ..., 5.836899, 5.549273, 6.050253],
       [1.589104, 2.051823, 1.479935, ..., 5.849903, 5.876003, 5.855825],
       [1.357033, 1.710564, 1.29594 , ..., 5.891077, 6.22814 , 5.934219],...
```

chengzhuzhang commented:
@tomvothecoder Thank you for the nice work so far. I walked through other sets and noticing problems that we should investigate and try to fix. 

[viewer/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/viewer/): for taylor diagram: results of TREFHT land is off, this should related to different masking.
[polar/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/) The difference plots have problems at the edge.
[meridional_mean_2d/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/meridional_mean_2d/): Max of middle panel = that of upper panel, and rmse is very different compare to original version
[zonal_mean_2d/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_2d/) and [zonal_mean_2d_stratosphere/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_2d_stratosphere/): metrics problem: Max of middle panel = that of upper panel, and rmse is always 0, mean is nan.

Cosmetic. 
[aerosol_aeronet/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/aerosol_aeronet/): Lat, lon labels diff
[zonal_mean_xy/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/): y axis name uses variable name instead of long name compared to old version, for ERA5-Z3 and T

chengzhuzhang commented:
> @chengzhuzhang For the `OMI-MLS-TCO-ANN-60S60N` variable, we are getting mismatching shapes for the `diff.nc` file.
> **Any suggestions on what we should do here? Related to #739.**

I checked the diff figures, despite the resolution difference, the metrics and results are very close. I think we can address #739 at a later time.

tomvothecoder commented:
> > @chengzhuzhang For the `OMI-MLS-TCO-ANN-60S60N` variable, we are getting mismatching shapes for the `diff.nc` file.
> > **Any suggestions on what we should do here? Related to #739.**
> 
> I checked the diff figures, despite the resolution difference, the metrics and results are very close. I think we can address #739 at a later time.

Sounds good. I've moved this info over to #739.

chengzhuzhang commented:
Updates on remaining issues:

[polar/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/) The difference plots have problems at the edge. example: [new](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-polar_N.png_actual.png) vs [old](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-polar_N.png_expected.png). The polar driver is to first subset data, and then regrid, and plot. The re-grided data has abnormal values at the latitude edge. Shown in below plotted regrided data for the reference in the middle panel. 
![image](https://github.com/user-attachments/assets/beb3216f-2c93-4633-b1e7-bef9c7a778e5)

[zonal_mean_2d/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_2d/) and [zonal_mean_2d_stratosphere/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_2d_stratosphere/): metrics problem: mean of ref is `nan` for relative difference plots.

Cosmetic.
[aerosol_aeronet/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/aerosol_aeronet/): Lat, lon labels diff
[zonal_mean_xy/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/): y axis name uses variable name instead of long name compared to old version, for ERA5-Z3 and T

chengzhuzhang commented on e3sm_diags/driver/meridional_mean_2d_driver.py:
```diff
@@ -218,7 +218,7 @@ def _create_metrics_dict(
     metrics_dict["ref"] = {
         "min": ds_ref[var_key].min().item(),
-        "max": ds_test[var_key].max().item(),
+        "max": ds_ref[var_key].max().item(),
```
fix a typo.

chengzhuzhang commented on e3sm_diags/driver/polar_driver.py:
```diff
@@ -279,7 +279,7 @@ def _create_metrics_dict(
     metrics_dict["ref"] = {
         "min": ds_ref[var_key].min().item(),
-        "max": ds_test[var_key].max().item(),
+        "max": ds_ref[var_key].max().item(),
```
fix a typo.

chengzhuzhang commented on e3sm_diags/driver/zonal_mean_2d_driver.py:
```diff
@@ -219,7 +219,7 @@ def _get_avg_for_regridded_datasets(
         ds_diff_rg_avg[var_key].attrs["units"] = "%"
 
-    return ds_test_rg_avg, ds_test_rg_avg, ds_diff_rg_avg
+    return ds_test_rg_avg, ds_ref_rg_avg, ds_diff_rg_avg
```
fix a typo.

chengzhuzhang commented on e3sm_diags/driver/zonal_mean_2d_driver.py:
```diff
@@ -262,7 +262,7 @@ def _create_metrics_dict(
     metrics_dict["ref"] = {
         "min": ds_ref[var_key].min().item(),
-        "max": ds_test[var_key].max().item(),
+        "max": ds_ref[var_key].max().item(),
```
fix a typo

chengzhuzhang commented on e3sm_diags/plot/polar_plot.py:
```diff
@@ -209,6 +209,7 @@ def _add_colormap(
         subplot_num,
         PANEL_CFG,
         metrics,
+        fontsize=PLOT_SECONDARY_TITLE,
```
Use same fontsize as v2 version to reduce image diffs.

chengzhuzhang commented:
@tomvothecoder I think we have fixed some problems that can significantly reduce the amount of image diff. Are you still interested to address one of those cosmetic inconsistency in https://github.com/E3SM-Project/e3sm_diags/pull/931#issuecomment-2652403024 for a new rc? Or we can merge and release a rc first for e3sm_unified testing, and address remaining issues on the side..

tomvothecoder commented on e3sm_diags/plot/aerosol_aeronet_plot.py:
```diff
@@ -108,7 +108,4 @@ def plot(
 
     plt.loglog(ref_site_arr, test_site_arr, "kx", markersize=3.0, mfc="none")
 
-    # Configure legend.
-    plt.legend(frameon=False, prop={"size": 5})
-
```
Removed unused `plt.legend()` that produces a UserWarning because there is no label for the data.

tomvothecoder commented on e3sm_diags/plot/utils.py:
```diff
@@ -53,6 +53,7 @@
     "meridional_mean_2d",
     "streamflow",
     "tc_analysis",
+    "aerosol_aeronet",
```
This fixes the `aerosol_aeronet` X and Y axes labels.

Example: https://web.lcrc.anl.gov/public/e3sm/cdat-migration-fy24/25-02-13-branch-930-zppy-diffs-aerosol/viewer/aerosol_aeronet/aeronet/aodabs-global-aeronet/ann.html

tomvothecoder reviewed (commented):
My aerosol_aeronet X and Y axes label fixes

tomvothecoder commented:
> @tomvothecoder I think we have fixed some problems that can significantly reduce the amount of image diff. Are you still interested to address one of those cosmetic inconsistency in [#931 (comment)](https://github.com/E3SM-Project/e3sm_diags/pull/931#issuecomment-2652403024) for a new rc? Or we can merge and release a rc first for e3sm_unified testing, and address remaining issues on the side..

I just fixed one of the two cosmetic issues. I'm working on the last when, then will merge when ready. I'll release v3.0.0rc3 subsequently and notify Xylar for E3SM Unified and Ryan for `zppy` testing.

chengzhuzhang commented:
> I just fixed one of the two cosmetic issues. I'm working on the last when, then will merge when ready. I'll release v3.0.0rc3 subsequently and notify Xylar for E3SM Unified and Ryan for `zppy` testing.

Sounds great. Thank you!

tomvothecoder commented:
@chengzhuzhang I noticed some `ceres_ebaf_toa` plots in `zonal_mean_xy` for "expected" (v2.12.1) look odd compared the the "actual" (v3.0.0rc2). Is the "expected" correct?

- Example [expected](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/CERES-EBAF-TOA-v4.1/ceres_ebaf_toa_v4.1-ALBEDO-JJA-global.png_expected.png) (1e-6 axes)
- Example [actual](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/CERES-EBAF-TOA-v4.1/ceres_ebaf_toa_v4.1-ALBEDO-JJA-global.png_actual.png)

chengzhuzhang commented:
> @chengzhuzhang I noticed the `ceres_ebaf_toa_v4.1-ALBEDO-JJA` plots in `zonal_mean_xy` for "expected" (v2.12.1) look odd compared the the "actual" (v3.0.0rc2). Is the "expected" correct?
> 
>     * Example [expected](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/CERES-EBAF-TOA-v4.1/ceres_ebaf_toa_v4.1-ALBEDO-JJA-global.png_expected.png) (1e-6 axes)
> 
>     * Example [actual](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/CERES-EBAF-TOA-v4.1/ceres_ebaf_toa_v4.1-ALBEDO-JJA-global.png_actual.png)

Nice catch! I think this has been seen before and is more or less expected. During the boreal summer, the high-latitude Southern Hemisphere receives very little incoming radiation. Albedo is defined as the ratio of reflected radiation to incoming radiation, which resulted in an extremely high albedo value. To address this, we had to exclude the high-latitude band from the lat-lon map, but this adjustment was not applied in the zonal mean line plot.

tomvothecoder commented on e3sm_diags/driver/utils/regrid.py:
```diff
@@ -628,6 +628,8 @@ def _hybrid_to_plevs(
     # Vertical regriding sets the units to "mb", but the original units
     # should be preserved.
     result[var_key].attrs["units"] = ds[var_key].attrs["units"]
+    result[var_key].attrs["standard_name"] = ds[var_key].attrs["standard_name"]
+    result[var_key].attrs["long_name"] = ds[var_key].attrs["long_name"]
```
Update `_hybrid_to_plevs()` to keep `long_name` and `standard_name` attrs

tomvothecoder commented on e3sm_diags/plot/zonal_mean_xy_plot.py:
```diff
@@ -50,7 +50,7 @@ def plot(
     lat_test = xc.get_dim_coords(da_test, axis="Y")
     lat_ref = xc.get_dim_coords(da_ref, axis="Y")
     lat_diff = xc.get_dim_coords(da_diff, axis="Y")
-    long_name = parameter.viewer_descr[parameter.var_id]
```
Fix `zonal_mean_xy_plot.py` to use correct `long_name` attribute

tomvothecoder reviewed (commented):
My fixes for `zonal_mean_xy` Y label to use `long_name` attribute.

Example: https://web.lcrc.anl.gov/public/e3sm/cdat-migration-fy24/25-02-13-branch-930-zonal-mean-xy/viewer/zonal_mean_xy/index.html

tomvothecoder commented on e3sm_diags/driver/zonal_mean_2d_driver.py:
```diff
@@ -216,7 +216,9 @@ def _get_avg_for_regridded_datasets(
 
     if parameter.diff_type == "relative":
-        ds_diff_rg_avg = ds_diff_rg_avg / ds_ref_rg_avg * 100.0
+        ds_diff_rg_avg[var_key] = (
+            ds_diff_rg_avg[var_key] / ds_ref_rg_avg[var_key] * 100.0
+        )
```
Fixes relative diff plots displaying mean `np.nan`.

Example: https://web.lcrc.anl.gov/public/e3sm/cdat-migration-fy24/25-02-13-branch-930-zonal-mean-2d-mean-fix/viewer/zonal_mean_2d/era5-relative-difference/q-global-era5/ann.html

tomvothecoder replied:
This issue was caused by performing arithmetic on the entire regridded and spatially averaged diffs dataset, which also affected other variables such as bounds (resulting in 0 weighting and `np.nan` mean).

tomvothecoder commented:
> Updates on remaining issues:
> 
>     * [ ]  [polar/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/) The difference plots have problems at the edge. example: [new](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-polar_N.png_actual.png) vs [old](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-polar_N.png_expected.png). The polar driver is to first subset data, and then regrid, and plot. The re-grided data has abnormal values at the latitude edge. Shown in below plotted regrided data for the reference in the middle panel.
>       ![image](https://private-user-images.githubusercontent.com/13056557/412217981-beb3216f-2c93-4633-b1e7-bef9c7a778e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0OTIwODUsIm5iZiI6MTczOTQ5MTc4NSwicGF0aCI6Ii8xMzA1NjU1Ny80MTIyMTc5ODEtYmViMzIxNmYtMmM5My00NjMzLWIxZTctYmVmOWM3YTc3OGU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE0VDAwMDk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwNGIzMmFhOThkODdhODI0OGQ2OWVmNTQyYTRiZWMwMDU5NTc4N2Q5NThhOTM0OTFlMmU4OGYxZWQwNWVhMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9KUzp_8BU6TrP50xErrw1e7yhcJiqbf2tEuDNmfXTLQ)
>       The native lat-lon grid doesn't show this pattern (mid panel):
>       ![image](https://private-user-images.githubusercontent.com/13056557/412218940-b80bbd1b-bea6-4805-8a1c-7db652128755.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0OTIwODUsIm5iZiI6MTczOTQ5MTc4NSwicGF0aCI6Ii8xMzA1NjU1Ny80MTIyMTg5NDAtYjgwYmJkMWItYmVhNi00ODA1LThhMWMtN2RiNjUyMTI4NzU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE0VDAwMDk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0YjFmMjYwOTJmMzBhYmJjMjY4ZmIzNTRjOTlkMTEzZjhhZDczYWIxMmY4ZDJjMWFjZDkxN2Q0NmUzYTU0N2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DhDwujWZMFmcKJmL6aPjyTbq3eTX8ovZ6eOjKiAJJAQ)
> 
>     * [x]  [zonal_mean_2d/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_2d/) and [zonal_mean_2d_stratosphere/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_2d_stratosphere/): metrics problem: mean of ref is `nan` for relative difference plots.
> 
> 
> Cosmetic.
> 
>     * [x]  [aerosol_aeronet/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/aerosol_aeronet/): Lat, lon labels diff
> 
>     * [x]  [zonal_mean_xy/](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/zonal_mean_xy/): y axis name uses variable name instead of long name compared to old version, for ERA5-Z3 and T

@chengzhuzhang I am going to try to address the last remaining issue from this list tomorrow, which is the polar one.
Then I'll make the release when ready.

tomvothecoder commented:
> Nice catch! I think this has been seen before and is more or less expected. During the boreal summer, the high-latitude Southern Hemisphere receives very little incoming radiation. Albedo is defined as the ratio of reflected radiation to incoming radiation, which resulted in an extremely high albedo value. To address this, we had to exclude the high-latitude band from the lat-lon map, but this adjustment was not applied in the zonal mean line plot.

Can you open a separate GitHub issue for us to address this later?

tomvothecoder commented:
> * The difference plots have problems at the edge. example: [new](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-polar_N.png_actual.png) vs [old](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_weekly_comprehensive_v2_www/test_pr651_both_commits_20250117/v2.LR.historical_0201/image_check_failures_comprehensive_v2/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1982-1983/polar/GPCP_v3.2/GPCP_v3.2-PRECT-ANN-polar_N.png_expected.png). The polar driver is to first subset data, and then regrid, and plot. The re-grided data has abnormal values at the latitude edge. Shown in below plotted regrided data for the reference in the middle panel.

I'm currently debugging this issue and think I found a potential bug/undesired behavior with xESMF. This requires further debugging in a separate PR. I'm going to merge this PR after my final self-review to proceed with v3.0.0rc3. 

### Potential Bug in xESMF

I found the test and reference datasets have different `bounds` dimension name (test has `nbnd`, ref has `bnd`). Since `bounds` dimension names don't align, xESMF seems to produce different results with regridding (maybe different interpretation of grid cells?). This contributes to larger floating point differences compared to the CDAT code. 

### Possible Workaround/Solution
However, if I drop those pre-existing bounds and do the regridding, xESMF is much closer to the CDAT code. I still noticed the abnormal data the edge of the plot though.

Example script below:
```python
"""
This script demonstrates the difference between regridding with and without
pre-existing lat/lon bounds.

It uses two datasets, one with lat/lon bounds and one without. The script
finds that the regridded arrays are not exactly the same, even though the
input arrays are the same.

The name of the bounds dimensions is different in the two datasets, which
causes the regridding to be different.
"""
import xcdat as xc
import numpy as np

ds_a = xc.open_dataset("test.nc")
ds_b = xc.open_dataset("ref.nc")

# 1. Check the bounds dimension names
print("Bounds in ds_a:", ds_a["lon_bnds"].dims, ds_a["lat_bnds"].dims)
print("Bounds in ds_b:", ds_b["lon_bnds"].dims, ds_b["lat_bnds"].dims)

"""
Bounds in ds_a: ('lon', 'nbnd') ('lat', 'nbnd')
Bounds in ds_b: ('lon', 'bnds') ('lat', 'bnds')

Notice how the bounds dimensions are different ('nbnd' vs. 'bnd'.). This
seems to affect results when regridding.
"""

# 2. Regrid with pre-existing lat/lon bounds. -- This one produces larger diffs compared to CDAT
output_grid = ds_a.regridder.grid
prect_a = ds_b.regridder.horizontal("PRECT", output_grid, tool="xesmf", method="conservative")["PRECT"]

# 3. Regrid without pre-existng bounds. -- This one is close to CDAT.
ds_a_no_bnds = ds_a.drop_vars(["lon_bnds", "lat_bnds"])
ds_b_no_bnds = ds_b.drop_vars(["lon_bnds", "lat_bnds"])
output_grid_no_bnds = ds_a_no_bnds.regridder.grid
prect_b = ds_b_no_bnds.regridder.horizontal("PRECT", output_grid_no_bnds, tool="xesmf", method="conservative")["PRECT"]

# 4. Print the stats -- they are "relatively" close but still different (which we shouldn't expect).
def print_stats(arr1, arr2, label1="Array 1", label2="Array 2"):
    stats = {
        "Min": (np.min(arr1), np.min(arr2)),
        "Max": (np.max(arr1), np.max(arr2)),
        "Mean": (np.mean(arr1), np.mean(arr2)),
        "Sum": (np.sum(arr1), np.sum(arr2)),
        "Std": (np.std(arr1), np.std(arr2)),
    }

    print(f"\n{'Stat':<10} {label1:<15} {label2:<15}")
    print("-" * 40)
    for stat, values in stats.items():
        print(f"{stat:<10} {values[0]:<15.6f} {values[1]:<15.6f}")

print_stats(prect_a.values, prect_b.values, "PRECT (with bounds)", "PRECT (no bounds)")

"""
Stat       PRECT (with bounds) PRECT (no bounds)
----------------------------------------
Min        0.165452        0.164364
Max        9.341431        10.046874
Mean       1.394370        1.398573
Sum        20078.931641    20139.449219
Std        1.016806        1.039437
"""

# 5. Compare arrays for floating point differences -- notice how they don't align
try:
    np.testing.assert_allclose(prect_a.values, prect_b.values, rtol=1e-5, atol=0)
except AssertionError as e:
    print("Arrays are not within rtol 1e-5")
    print(e)
else:
    print("Arrays are within rtol 1e-5")

"""
Arrays are not within rtol 1e-5

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 14392 / 14400 (99.9%)
Max absolute difference among violations: 3.6998303
Max relative difference among violations: 1.2256833
 ACTUAL: array([[1.123808, 1.435736, 1.278921, ..., 1.565927, 1.16721 , 1.126871],
       [2.288256, 2.331007, 2.508708, ..., 2.673229, 2.40844 , 2.271105],
       [1.978262, 2.111453, 1.774192, ..., 2.471611, 2.190455, 2.001195],...
 DESIRED: array([[2.356618, 2.652923, 2.578529, ..., 2.912792, 2.402574, 2.316819],
       [2.049907, 2.211725, 2.076397, ..., 2.646537, 2.309698, 2.113865],
       [1.979461, 2.010734, 1.869982, ..., 2.347431, 2.147887, 1.941646],...

"""
```
